miner/remote_proposals.go

```diff
@@ -0,0 +1,219 @@
+		pb.signers.signers[sig.NodeID()] = &signerSession{
+			signer: sig,
+			log:    pb.logger.With(zap.String("signer", sig.NodeID().ShortString())),
+		}
+	}
+}
+
+// Start the loop that listens to layers and build proposals.
+func (pb *RemoteProposalBuilder) Run(ctx context.Context) error {
+	current := pb.clock.CurrentLayer()
+	next := current + 1
+	pb.logger.Info("started", zap.Inline(&pb.cfg), zap.Uint32("next", next.Uint32()))
+	var eg errgroup.Group
+	prepareDisabled := pb.cfg.activeSet.Tries == 0 || pb.cfg.activeSet.RetryInterval == 0
+	if prepareDisabled {
+		pb.logger.Warn("activeset will not be prepared in advance")
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			eg.Wait()
+			return nil
+		case <-pb.clock.AwaitLayer(next):
+			current := pb.clock.CurrentLayer()
+			if current.Before(next) {
+				pb.logger.Info("time sync detected, realigning ProposalBuilder",
+					zap.Uint32("current", current.Uint32()),
+					zap.Uint32("next", next.Uint32()),
+				)
+				continue
+			}
+			next = current.Add(1)
+			ctx := log.WithNewSessionID(ctx)
+
+			if current <= types.GetEffectiveGenesis() {
+				continue
+			}
+			if err := pb.build(ctx, current); err != nil {
+				pb.logger.Warn("failed to build proposal",
+					log.ZContext(ctx),
+					zap.Uint32("lid", current.Uint32()),
+					zap.Error(err),
+				)
+			}
+
+			pb.clean(current)
+		}
+	}
+}
+
+func (pb *RemoteProposalBuilder) clean(layer types.LayerID) {
+	var vals []types.EpochID
+	lim := layer.GetEpoch() - 2
+	for k := range pb.epochEligibilities {
+		if k <= lim {
+			vals = append(vals, k)
+		}
+	}
+	for _, v := range vals {
+		delete(pb.epochEligibilities, v)
+	}
+}
+
+func (pb *RemoteProposalBuilder) build(ctx context.Context, layer types.LayerID) error {
+	pb.signers.mu.Lock()
+	signers := maps.Values(pb.signers.signers)
+	pb.signers.mu.Unlock()
+	for _, signer := range signers {
+		proposal, nonce, err := pb.nodeSvc.Proposal(ctx, layer, signer.signer.NodeID())
+		if err != nil {
+			pb.logger.Error("get partial proposal", zap.Error(err))
+			continue
+		}
+		if proposal == nil {
+			// this node signer isn't eligible this epoch, continue
+			pb.logger.Info("node not eligible on this layer. will try later")
+			continue
+		}
+		bcn, err := pb.nodeSvc.Beacon(ctx, layer.GetEpoch())
+		if err != nil {
+			pb.logger.Error("get beacon", zap.Error(err))
+			continue
+		}
+		var (
+			elig uint32
+			ok   bool
+		)
+		if proposal.Ballot.EpochData != nil {
+			elig, ok = pb.epochEligibilities[layer.GetEpoch()]
+			if !ok {
+				pb.epochEligibilities[layer.GetEpoch()] = proposal.Ballot.EpochData.EligibilityCount
+				elig = proposal.Ballot.EpochData.EligibilityCount
+			}
+		} else {
+			elig, ok = pb.epochEligibilities[layer.GetEpoch()]
+			if !ok {
+				panic("missing epoch eligibilities")
+			}
+		}
+
+		proofs := calcEligibilityProofs(
+			signer.signer.VRFSigner(),
+			layer.GetEpoch(),
+			bcn,
+			types.VRFPostIndex(nonce),
+			elig,
+			pb.cfg.layersPerEpoch,
+		)
```

Comment:
I think the proofs could be calculated once and cached for the rest of the epoch (they don't change), wdyt?

Reply:
yes and done